File: Python/boj/Silver/11053.py
import sys

if __name__ == "__main__":
    try:
        sys.stdin = open("input.txt", "r")
    except FileNotFoundError:
        pass

    N = int(input())
    numbers = list(map(int, input().split(" ")))
    dp = [0] * N

    dp[0] = 1

    def findIndex(i):
        index = -1
        max = 0

        for j in range(0, i):
            if numbers[j] < numbers[i] and max < dp[j]:
                max = dp[j]
                index = j

        return index if index != -1 else i

    for i in range(1, N):
        index = findIndex(i)
        dp[i] = dp[index] + 1

    print(max(dp))

File: JavaScript/boj/Silver/11053.js
const fs = require('fs');
const filePath = process.platform === 'linux' ? '/dev/stdin' : './input.txt';
const input = fs.readFileSync(filePath).toString().trim().split('\n');

const N = +input[0];
const numbers = input[1].split(' ').map(Number);
const dp = Array.from({ length: N }, () => 1);

for (let i = 1; i < N; i++) {
  let max = 0;
  let index = -1;

  for (let j = 0; j < i; j++) {
    if (numbers[j] < numbers[i] && max < dp[j]) {
      max = dp[j];
      index = j;
    }
  }

  if (index != -1) dp[i] = dp[index] + 1;
}

console.log(Math.max(...dp));
